File: scripts/indicators.py
#!/usr/bin/env python3
"""
Fetch Yahoo Finance OHLCV for BVB symbols and compute technical indicators.

Standard library only — runs anywhere Python 3.8+ is available.

Usage:
    python3 scripts/indicators.py SNG TLV BRD
    python3 scripts/indicators.py --symbols-file universe.txt
    python3 scripts/indicators.py --format=json SNG H2O   # machine-readable
    python3 scripts/indicators.py --format=table SNG H2O  # human-readable

Exit codes:
    0 — all requested symbols returned data
    1 — at least one symbol had no data (partial output still written)
    2 — fatal (bad args, network totally dead, etc.)
"""

import argparse
import json
import sys
import urllib.parse
import urllib.request
from typing import Any

YAHOO_CHART = "https://query1.finance.yahoo.com/v8/finance/chart/{sym}?interval=1d&range=100d"
STOOQ_CSV = "https://stooq.com/q/d/l/?s={sym}.ro&i=d"

UA = "Mozilla/5.0"


def fetch_yahoo(symbol: str) -> dict[str, Any] | None:
    url = YAHOO_CHART.format(sym=urllib.parse.quote(symbol))
    req = urllib.request.Request(url, headers={"User-Agent": UA})
    try:
        with urllib.request.urlopen(req, timeout=15) as resp:
            data = json.loads(resp.read().decode("utf-8"))
    except Exception as e:
        print(f"[warn] yahoo fetch failed for {symbol}: {e}", file=sys.stderr)
        return None
    result = (data.get("chart") or {}).get("result") or []
    if not result:
        return None
    return result[0]


def rsi_wilder(closes: list[float], period: int = 14) -> float | None:
    if len(closes) < period + 1:
        return None
    gains = losses = 0.0
    for i in range(-period, 0):
        delta = closes[i] - closes[i - 1]
        if delta > 0:
            gains += delta
        else:
            losses += -delta
    avg_gain = gains / period
    avg_loss = losses / period
    if avg_loss == 0:
        return 100.0
    rs = avg_gain / avg_loss
    return 100.0 - 100.0 / (1.0 + rs)


def sma(values: list[float], window: int) -> float | None:
    if len(values) < window:
        return None
    return sum(values[-window:]) / window


def atr_pct(highs: list[float], lows: list[float], closes: list[float], period: int = 14) -> float | None:
    if len(closes) < period + 1:
        return None
    trs = []
    for i in range(-period, 0):
        h, l, pc = highs[i], lows[i], closes[i - 1]
        trs.append(max(h - l, abs(h - pc), abs(l - pc)))
    atr = sum(trs) / period
    return atr / closes[-1] * 100


def trend_label(price: float, sma20: float | None, sma50: float | None) -> str:
    if sma50 is None:
        return "unknown"
    if price > sma50 and (sma20 is None or sma20 > sma50):
        return "up"
    if price < sma50:
        return "down"
    return "range"


def compute(symbol: str) -> dict[str, Any] | None:
    raw = fetch_yahoo(symbol)
    if raw is None:
        return None
    meta = raw.get("meta") or {}
    quote = ((raw.get("indicators") or {}).get("quote") or [{}])[0]

    closes = [c for c in (quote.get("close") or []) if c is not None]
    highs = [h for h in (quote.get("high") or []) if h is not None]
    lows = [l for l in (quote.get("low") or []) if l is not None]
    vols = [v for v in (quote.get("volume") or []) if v is not None]

    if not closes:
        return None

    # align arrays — sometimes close/high/low have different None positions.
    # simplest safe approach: trim to common length from the right.
    n = min(len(closes), len(highs), len(lows))
    closes, highs, lows = closes[-n:], highs[-n:], lows[-n:]
    if len(vols) > n:
        vols = vols[-n:]

    price = meta.get("regularMarketPrice") or closes[-1]
    # chartPreviousClose is the close BEFORE the range window starts, not yesterday.
    # Use the second-to-last close in the quote array for true day-over-day change.
    prev_close = closes[-2] if len(closes) >= 2 else meta.get("chartPreviousClose")

    sma20 = sma(closes, 20)
    sma50 = sma(closes, 50)
    sma200 = sma(closes, 200)

    # volume ratio — today's volume vs 20d average
    avg_vol20 = (sum(vols[-20:]) / 20) if len(vols) >= 20 else None
    today_vol = vols[-1] if vols else None
    vol_ratio = (today_vol / avg_vol20) if (avg_vol20 and today_vol is not None) else None

    h20 = max(highs[-20:]) if len(highs) >= 20 else None
    l20 = min(lows[-20:]) if len(lows) >= 20 else None

    return {
        "symbol": symbol,
        "price": price,
        "prev_close": prev_close,
        "daily_change_pct": ((price - prev_close) / prev_close * 100) if (prev_close and price) else None,
        "rsi14": rsi_wilder(closes),
        "sma20": sma20,
        "sma50": sma50,
        "sma200": sma200,
        "atr14_pct": atr_pct(highs, lows, closes),
        "volume_today": today_vol,
        "volume_avg20": avg_vol20,
        "volume_ratio": vol_ratio,
        "high_20d": h20,
        "low_20d": l20,
        "high_52w": meta.get("fiftyTwoWeekHigh"),
        "low_52w": meta.get("fiftyTwoWeekLow"),
        "trend": trend_label(price, sma20, sma50),
        "currency": meta.get("currency"),
        "exchange": meta.get("exchangeName"),
        "n_bars": len(closes),
    }


def format_table(rows: list[dict[str, Any]]) -> str:
    header = f"{'SYM':<6} {'PRICE':>9} {'CHG%':>6} {'RSI14':>6} {'SMA20':>9} {'SMA50':>9} {'TREND':>7} {'VOL×':>5} {'20dH':>9} {'20dL':>9} {'52wH':>9} {'ATR%':>5}"
    lines = [header, "-" * len(header)]
    for r in rows:
        if not r:
            continue
        def fmt(v, w, d=2):
            if v is None:
                return " " * (w - 1) + "-"
            return f"{v:>{w}.{d}f}"
        lines.append(
            f"{r['symbol']:<6} "
            f"{fmt(r['price'], 9, 3)} "
            f"{fmt(r.get('daily_change_pct'), 6, 2)} "
            f"{fmt(r['rsi14'], 6, 1)} "
            f"{fmt(r['sma20'], 9, 3)} "
            f"{fmt(r['sma50'], 9, 3)} "
            f"{r['trend']:>7} "
            f"{fmt(r['volume_ratio'], 5, 1)} "
            f"{fmt(r['high_20d'], 9, 3)} "
            f"{fmt(r['low_20d'], 9, 3)} "
            f"{fmt(r['high_52w'], 9, 3)} "
            f"{fmt(r['atr14_pct'], 5, 1)}"
        )
    return "\n".join(lines)


def main() -> int:
    p = argparse.ArgumentParser(description="Compute BVB technical indicators from Yahoo Finance")
    p.add_argument("symbols", nargs="*", help="BVB tickers (e.g. SNG TLV BRD). Appends '.RO' automatically.")
    p.add_argument("--symbols-file", help="Read tickers from file, one per line")
    p.add_argument("--format", choices=["json", "table"], default="json", help="Output format")
    p.add_argument("--suffix", default=".RO", help="Exchange suffix to append (default: .RO)")
    args = p.parse_args()

    symbols: list[str] = list(args.symbols)
    if args.symbols_file:
        with open(args.symbols_file) as f:
            symbols.extend(line.strip() for line in f if line.strip() and not line.startswith("#"))

    if not symbols:
        print("error: no symbols given (positional args or --symbols-file)", file=sys.stderr)
        return 2

    rows: list[dict[str, Any] | None] = []
    for sym in symbols:
        yahoo_sym = sym if "." in sym else f"{sym}{args.suffix}"
        r = compute(yahoo_sym)
        if r:
            r["symbol"] = sym  # report the bare ticker for downstream use
        rows.append(r)

    any_missing = any(r is None for r in rows)

    if args.format == "json":
        print(json.dumps([r for r in rows if r], indent=2))
    else:
        print(format_table([r for r in rows if r]))

    missing = [sym for sym, r in zip(symbols, rows) if r is None]
    if missing:
        print(f"\n[warn] no data for: {', '.join(missing)}", file=sys.stderr)

    return 1 if any_missing else 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/indicators.mjs
#!/usr/bin/env node
/**
 * Fetch Yahoo Finance OHLCV for BVB symbols and compute technical indicators.
 *
 * Node 18+ stdlib only (uses built-in fetch). No runtime deps.
 *
 * Usage:
 *     node scripts/indicators.mjs SNG TLV BRD
 *     node scripts/indicators.mjs --symbols-file universe.txt
 *     node scripts/indicators.mjs --format=json SNG H2O     # machine-readable
 *     node scripts/indicators.mjs --format=table SNG H2O    # human-readable
 *
 * Exit codes:
 *     0 — all requested symbols returned data
 *     1 — at least one symbol had no data (partial output still written)
 *     2 — fatal (bad args, network totally dead, etc.)
 */

import fs from 'node:fs';

const YAHOO_CHART = (sym) =>
  `https://query1.finance.yahoo.com/v8/finance/chart/${encodeURIComponent(sym)}?interval=1d&range=100d`;
const UA = 'Mozilla/5.0';

async function fetchYahoo(symbol) {
  const url = YAHOO_CHART(symbol);
  try {
    const resp = await fetch(url, {
      headers: { 'User-Agent': UA },
      signal: AbortSignal.timeout(15000),
    });
    if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
    const data = await resp.json();
    const result = data?.chart?.result || [];
    return result[0] || null;
  } catch (e) {
    process.stderr.write(`[warn] yahoo fetch failed for ${symbol}: ${e.message}\n`);
    return null;
  }
}

// Wilder's RSI — matches sim_executor.py exactly on the last `period+1` closes.
function rsiWilder(closes, period = 14) {
  if (closes.length < period + 1) return null;
  let gains = 0, losses = 0;
  // Python range(-period, 0) → iterate indices len-period .. len-1
  for (let i = closes.length - period; i < closes.length; i++) {
    const delta = closes[i] - closes[i - 1];
    if (delta > 0) gains += delta;
    else losses += -delta;
  }
  const avgGain = gains / period;
  const avgLoss = losses / period;
  if (avgLoss === 0) return 100.0;
  const rs = avgGain / avgLoss;
  return 100.0 - 100.0 / (1.0 + rs);
}

function sma(values, window) {
  if (values.length < window) return null;
  let s = 0;
  for (let i = values.length - window; i < values.length; i++) s += values[i];
  return s / window;
}

function atrPct(highs, lows, closes, period = 14) {
  if (closes.length < period + 1) return null;
  const trs = [];
  for (let i = closes.length - period; i < closes.length; i++) {
    const h = highs[i], l = lows[i], pc = closes[i - 1];
    trs.push(Math.max(h - l, Math.abs(h - pc), Math.abs(l - pc)));
  }
  const atr = trs.reduce((a, b) => a + b, 0) / period;
  return (atr / closes[closes.length - 1]) * 100;
}

function trendLabel(price, sma20, sma50) {
  if (sma50 == null) return 'unknown';
  if (price > sma50 && (sma20 == null || sma20 > sma50)) return 'up';
  if (price < sma50) return 'down';
  return 'range';
}

async function compute(symbol) {
  const raw = await fetchYahoo(symbol);
  if (!raw) return null;
  const meta = raw.meta || {};
  const quote = raw?.indicators?.quote?.[0] || {};

  const closes = (quote.close || []).filter((c) => c !== null && c !== undefined);
  const highs  = (quote.high  || []).filter((h) => h !== null && h !== undefined);
  const lows   = (quote.low   || []).filter((l) => l !== null && l !== undefined);
  let vols     = (quote.volume || []).filter((v) => v !== null && v !== undefined);

  if (!closes.length) return null;

  // Align arrays — trim to common length from the right.
  const n = Math.min(closes.length, highs.length, lows.length);
  const cs = closes.slice(-n);
  const hs = highs.slice(-n);
  const ls = lows.slice(-n);
  if (vols.length > n) vols = vols.slice(-n);

  const price = meta.regularMarketPrice ?? cs[cs.length - 1];
  // chartPreviousClose is pre-range, not yesterday. Use second-to-last close.
  const prevClose = cs.length >= 2 ? cs[cs.length - 2] : meta.chartPreviousClose;

  const s20 = sma(cs, 20);
  const s50 = sma(cs, 50);
  const s200 = sma(cs, 200);

  const avgVol20 = vols.length >= 20
    ? vols.slice(-20).reduce((a, b) => a + b, 0) / 20
    : null;
  const todayVol = vols.length ? vols[vols.length - 1] : null;
  const volRatio = (avgVol20 && todayVol != null) ? (todayVol / avgVol20) : null;

  const h20 = hs.length >= 20 ? Math.max(...hs.slice(-20)) : null;
  const l20 = ls.length >= 20 ? Math.min(...ls.slice(-20)) : null;

  return {
    symbol,
    price,
    prev_close: prevClose,
    daily_change_pct: (prevClose && price) ? ((price - prevClose) / prevClose) * 100 : null,
    rsi14: rsiWilder(cs),
    sma20: s20,
    sma50: s50,
    sma200: s200,
    atr14_pct: atrPct(hs, ls, cs),
    volume_today: todayVol,
    volume_avg20: avgVol20,
    volume_ratio: volRatio,
    high_20d: h20,
    low_20d: l20,
    high_52w: meta.fiftyTwoWeekHigh ?? null,
    low_52w: meta.fiftyTwoWeekLow ?? null,
    trend: trendLabel(price, s20, s50),
    currency: meta.currency ?? null,
    exchange: meta.exchangeName ?? null,
    n_bars: cs.length,
  };
}

function padLeft(str, w) { str = String(str); return str.length >= w ? str : ' '.repeat(w - str.length) + str; }
function padRight(str, w) { str = String(str); return str.length >= w ? str : str + ' '.repeat(w - str.length); }

function fmtNum(v, w, d = 2) {
  if (v == null) return ' '.repeat(w - 1) + '-';
  return padLeft(v.toFixed(d), w);
}

function formatTable(rows) {
  const header =
    `${padRight('SYM', 6)} ${padLeft('PRICE', 9)} ${padLeft('CHG%', 6)} ${padLeft('RSI14', 6)} ` +
    `${padLeft('SMA20', 9)} ${padLeft('SMA50', 9)} ${padLeft('TREND', 7)} ${padLeft('VOL×', 5)} ` +
    `${padLeft('20dH', 9)} ${padLeft('20dL', 9)} ${padLeft('52wH', 9)} ${padLeft('ATR%', 5)}`;
  const lines = [header, '-'.repeat(header.length)];
  for (const r of rows) {
    if (!r) continue;
    lines.push(
      `${padRight(r.symbol, 6)} ` +
      `${fmtNum(r.price, 9, 3)} ` +
      `${fmtNum(r.daily_change_pct, 6, 2)} ` +
      `${fmtNum(r.rsi14, 6, 1)} ` +
      `${fmtNum(r.sma20, 9, 3)} ` +
      `${fmtNum(r.sma50, 9, 3)} ` +
      `${padLeft(r.trend, 7)} ` +
      `${fmtNum(r.volume_ratio, 5, 1)} ` +
      `${fmtNum(r.high_20d, 9, 3)} ` +
      `${fmtNum(r.low_20d, 9, 3)} ` +
      `${fmtNum(r.high_52w, 9, 3)} ` +
      `${fmtNum(r.atr14_pct, 5, 1)}`
    );
  }
  return lines.join('\n');
}

function parseArgs(argv) {
  const args = { symbols: [], symbolsFile: null, format: 'json', suffix: '.RO' };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === '--symbols-file') { args.symbolsFile = argv[++i]; }
    else if (a.startsWith('--symbols-file=')) { args.symbolsFile = a.split('=', 2)[1]; }
    else if (a === '--format') { args.format = argv[++i]; }
    else if (a.startsWith('--format=')) { args.format = a.split('=', 2)[1]; }
    else if (a === '--suffix') { args.suffix = argv[++i]; }
    else if (a.startsWith('--suffix=')) { args.suffix = a.split('=', 2)[1]; }
    else if (a === '-h' || a === '--help') { args.help = true; }
    else { args.symbols.push(a); }
  }
  if (!['json', 'table'].includes(args.format)) {
    throw new Error(`--format must be json|table, got ${args.format}`);
  }
  return args;
}

async function main() {
  let args;
  try { args = parseArgs(process.argv.slice(2)); }
  catch (e) { process.stderr.write(`error: ${e.message}\n`); return 2; }

  if (args.help) {
    process.stdout.write(
      'Usage: node scripts/indicators.mjs [--format=json|table] [--suffix=.RO] [--symbols-file=FILE] SYM [SYM ...]\n'
    );
    return 0;
  }

  const symbols = [...args.symbols];
  if (args.symbolsFile) {
    const content = fs.readFileSync(args.symbolsFile, 'utf8');
    for (const line of content.split(/\r?\n/)) {
      const t = line.trim();
      if (t && !t.startsWith('#')) symbols.push(t);
    }
  }

  if (!symbols.length) {
    process.stderr.write('error: no symbols given (positional args or --symbols-file)\n');
    return 2;
  }

  const rows = [];
  for (const sym of symbols) {
    const yahooSym = sym.includes('.') ? sym : `${sym}${args.suffix}`;
    const r = await compute(yahooSym);
    if (r) r.symbol = sym;  // report the bare ticker for downstream use
    rows.push(r);
  }

  const anyMissing = rows.some((r) => r === null);
  const present = rows.filter((r) => r !== null);

  if (args.format === 'json') {
    process.stdout.write(JSON.stringify(present, null, 2) + '\n');
  } else {
    process.stdout.write(formatTable(present) + '\n');
  }

  const missing = symbols.filter((_, i) => rows[i] === null);
  if (missing.length) {
    process.stderr.write(`\n[warn] no data for: ${missing.join(', ')}\n`);
  }

  return anyMissing ? 1 : 0;
}

main().then(
  (code) => process.exit(code),
  (err) => { process.stderr.write(`FATAL: ${err.stack || err.message || err}\n`); process.exit(2); }
);
